src/Functions/Regexps.h

```diff
@@ -136,34 +137,36 @@ class Regexps
     ScratchPtr scratch;
 };
 
-class RegexpsConstructor
+class DeferredConstructedRegexps
 {
 public:
-    RegexpsConstructor() = default;
-
-    void setConstructor(std::function<Regexps()> constructor_) { constructor = std::move(constructor_); }
+    explicit DeferredConstructedRegexps(std::function<Regexps()> constructor_)
+        : constructor(std::move(constructor_))
+    {}
 
-    Regexps * operator()()
+    Regexps * get()
     {
         std::lock_guard lock(mutex);
-        if (regexp)
-            return &*regexp;
-        regexp = constructor();
-        return &*regexp;
+        if (regexps)
+            return &*regexps;
+        regexps = constructor();
+        return &*regexps;
```

Comment:
```
static std::optional<Regexps> regexps = [&]
{
  return constructor();
}();

return &*regexps;
```

no need for mutex or `regexps` field.

Reply:
I doesn't need to be optional.

Reply:
You propose a static local variable which is initialized just once according to C++11 rules. Yes, that would avoid the mutex and optional. The problem is that this would lead to one regexps instance for all DeferredConstructedRegexps objects. But we really need one regexps instance per DeferredConstructedRegexps object. So unfortunately, we need to retain the mutex + optional.

Reply:
ah yes, you're correct